src/sources/journald.rs

```diff
@@ -372,32 +293,134 @@ where
+type StartJournalctlFn = Box<
+    dyn Fn(
+            &Option<String>, // cursor
+        ) -> crate::Result<(BoxStream<'static, io::Result<Bytes>>, StopJournalctlFn)>
+        + Send
+        + Sync,
+>;
+
+type StopJournalctlFn = Box<dyn FnOnce() + Send>;
+
+fn start_journalctl(
+    path: &PathBuf,
+    current_boot_only: bool,
+    cursor: &Option<String>,
+) -> crate::Result<(BoxStream<'static, io::Result<Bytes>>, StopJournalctlFn)> {
+    let mut command = Command::new(path);
+    command.stdout(Stdio::piped());
+    command.arg("--follow");
+    command.arg("--all");
+    command.arg("--show-cursor");
+    command.arg("--output=json");
+
+    if current_boot_only {
+        command.arg("--boot");
+    }
+
+    if let Some(cursor) = cursor {
+        command.arg(format!("--after-cursor={}", cursor));
+    } else {
+        // journalctl --follow only outputs a few lines without a starting point
+        command.arg("--since=2000-01-01");
+    }
+
+    let mut child = command.spawn().context(JournalctlSpawn)?;
+
+    let stream = FramedRead::new(
+        child.stdout.take().unwrap(),
+        BytesDelimitedCodec::new(b'\n'),
+    )
+    .boxed();
+
+    let pid = Pid::from_raw(child.id() as i32);
+    let stop = Box::new(move || {
+        let _ = kill(pid, Signal::SIGTERM);
+    });
+
+    Ok((stream, stop))
+}
+
+fn create_event(record: Record) -> Event {
+    let mut log = LogEvent::from_iter(record);
+    // Convert some journald-specific field names into Vector standard ones.
+    if let Some(message) = log.remove(MESSAGE) {
+        log.insert(log_schema().message_key(), message);
+    }
+    if let Some(host) = log.remove(HOSTNAME) {
+        log.insert(log_schema().host_key(), host);
+    }
+    // Translate the timestamp, and so leave both old and new names.
+    if let Some(timestamp) = log
+        .get(&*SOURCE_TIMESTAMP)
+        .or_else(|| log.get(RECEIVED_TIMESTAMP))
+    {
+        if let Value::Bytes(timestamp) = timestamp {
+            if let Ok(timestamp) = String::from_utf8_lossy(timestamp).parse::<u64>() {
+                let timestamp = chrono::Utc.timestamp(
+                    (timestamp / 1_000_000) as i64,
+                    (timestamp % 1_000_000) as u32 * 1_000,
+                );
+                log.insert(log_schema().timestamp_key(), Value::Timestamp(timestamp));
             }
         }
     }
+    // Add source type
+    log.try_insert(log_schema().source_type_key(), Bytes::from("journald"));
+
+    log.into()
+}
+
+/// Map the given unit name into a valid systemd unit
+/// by appending ".service" if no extension is present.
+fn fixup_unit(unit: &str) -> String {
+    if unit.contains('.') {
+        unit.into()
+    } else {
+        format!("{}.service", unit)
+    }
 }
 
-fn decode_record(text: &str, remap: bool) -> Result<Record, JsonError> {
-    let mut record = serde_json::from_str::<JsonValue>(&text)?;
+fn decode_record(line: &[u8], remap: bool) -> Result<Record, JsonError> {
```

Comment:
Changed to bytes to avoid unnecessary clone (https://github.com/timberio/vector/pull/4243#discussion_r498610519).